packages/plugin-defender/src/hardhat/propose-upgrade.ts

```diff
@@ -0,0 +1,34 @@
+import '@openzeppelin/hardhat-upgrades/dist/type-extensions';
+import { getChainId, ValidationOptions } from '@openzeppelin/upgrades-core';
+import { AdminClient, ProposalResponse } from 'defender-admin-client';
+import type { ContractFactory } from 'ethers';
+import { FormatTypes } from 'ethers/lib/utils';
+import { HardhatRuntimeEnvironment } from 'hardhat/types';
+import { fromChainId } from './network';
+import { getProposalUrl } from './utils';
+
+export type ProposeUpgradeFunction = (
+  proxyAddress: string,
+  ImplFactory: ContractFactory,
+  opts?: ValidationOptions & { title?: string; description?: string },
+) => Promise<ProposalResponse & { url: string }>;
+
+export function makeProposeUpgrade(hre: HardhatRuntimeEnvironment): ProposeUpgradeFunction {
+  return async function proposeUpgrade(proxyAddress, ImplFactory, opts = {}) {
+    if (!hre.config.defender) {
+      throw new Error(`Missing Defender API key and secret in hardhat config`);
+    }
```

Comment:
Isn't this already checked in `extendConfig`?

Reply:
No, extendConfig just warns if the api key is not set. This way the user does not need to enter api key and secret if they run a buidler script that does not interact with defender.